src/Lean/Expr.lean

```diff
@@ -852,6 +852,33 @@ def getForallBody : Expr → Expr
   | forallE _ _ b .. => getForallBody b
   | e                => e
 
+def getForallBodyMaxDepth : (maxDepth : Nat) → Expr → Expr
+  | (n+1), forallE _ _ b _ => getForallBodyMaxDepth n b
+  | 0, e => e
+  | _, e => e
+
+/-- A binder is the `(a : α)` that appears in `(a : α) → β` or `fun (a : α) => B`.
+It consists of a user-friendly name, a type expression and the BinderInfo. -/
+structure Binder where
+  name : Name
+  type : Expr
+  info : BinderInfo
+
+/-- Given a sequence of nested foralls `(a₁ : α₁) → ... → (aₙ : αₙ) → _`,
+returns `[(a₁ : α₁), ..., (aₙ : αₙ)]`. -/
+def getForallBinders : Expr → List Binder
```

Comment:
This is a somewhat dangerous operation as it can expose loose bvars. I would suggest to merge it into `getForallBinderNames`, and perhaps to move all of these to the use site as they are quite usage specific.